porch/repository/pkg/git/git.go

```diff
@@ -643,3 +630,106 @@ func (r *gitRepository) getRepo() (string, error) {
+		return err
+	}
+
+	// Fetch
+	switch err := r.repo.Fetch(&gogit.FetchOptions{
+		RemoteName: originName,
+		Auth:       auth,
+		Tags:       gogit.AllTags,
+	}); err {
+	case nil: // OK
+	case gogit.NoErrAlreadyUpToDate:
+	case transport.ErrEmptyRemoteRepository:
+
+	default:
+		return fmt.Errorf("cannot fetch repository %s/%s: %w", r.namespace, r.name, err)
+	}
+
+	// Create tracking branches for remotes
+	refs, err := r.repo.References()
+	if err != nil {
+		return fmt.Errorf("cannot identify repository remote references in %s/%s: %w", r.namespace, r.name, err)
+	}
+
+	// Collect remote and local branches. Both maps are indexed by the local reference name.
+	// remote references (refs/remotes/origin/...) are transformed to `refs/heads/...` to have
+	// both maps use matching keys.
+	remoteBranches := map[string]*plumbing.Reference{}
+	localBranches := map[string]*plumbing.Reference{}
+	for {
+		ref, err := refs.Next()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			klog.Warningf("Skipping reference during iteration on error: %v", err)
+			continue
+		}
+
+		name := ref.Name().String()
+		switch {
+		case strings.HasPrefix(name, refRemoteBranchPrefix):
+			branch := strings.TrimPrefix(name, refRemoteBranchPrefix)
+			remoteBranches[refHeadsPrefix+branch] = ref
+
+		case strings.HasPrefix(name, refHeadsPrefix):
+			localBranches[name] = ref
+		}
+	}
+
+	// TODO: Explore use of automatic branch tracking to do this.
+	// There may be risks involved such as automated rebase which may lead to incorrect results on the package.
+	// One possibility is to replay package history ... something to consider in the future.
+	for name, remoteRef := range remoteBranches {
+		localRef, ok := localBranches[name]
+
+		if !ok {
+			localRef = plumbing.NewHashReference(plumbing.ReferenceName(name), remoteRef.Hash())
+			// local branch doesn't exist. create it
+			if err := r.repo.Storer.SetReference(localRef); err != nil {
+				klog.Warningf("Skipping failed local ref %q: %v", name, err)
+			}
+		} else if remoteRef.Hash() != localRef.Hash() {
+			remoteCommit, err := r.repo.CommitObject(remoteRef.Hash())
+			if err != nil {
+				klog.Warningf("Skipping unresolvable remote reference %s: %v", remoteRef, err)
+				continue
+			}
+			localCommit, err := r.repo.CommitObject(localRef.Hash())
+			if err != nil {
+				klog.Warningf("Overwriting unresolvable local reference %s: %v", localRef, err)
+				new := plumbing.NewHashReference(localRef.Name(), remoteCommit.Hash)
+				if err := r.repo.Storer.SetReference(new); err != nil {
+					klog.Errorf("Failed to set local reference: %s to %s", new.Name(), new.Hash())
+				}
+				continue
+			}
+
+			// If local commit is ancestor of remote, fast-forward local commit to match
+			ancestor, err := localCommit.IsAncestor(remoteCommit)
+			if err != nil {
+				klog.Warningf("Failed to determine whether %s is ancestor of %s: %v", localRef, remoteRef, err)
+			}
+
+			// TODO: Better conflict resolution policy
```

Comment:
So in my mind, the remote git store is the source of truth.  I guess the question is ... what happened here to cause the conflict?  I would imagine we would _always_ take the remote git store value, but there could well be a case I'm missing...

Reply:
race conditions can happen. Porch is processing package update; in parallel the remote repo gets new commit.

So long as Porch is the exclusive user of the Git repo all should be fine, but with other actors involved, we may see conflicts where draft branch is updated to 2 different, divergent, commits. May be mostly theoretical scenario.